### ISSUE TITLE
NFS conflict with composer

### ISSUE THREAD

**Issue by loiclavoie** [composer/composer#6468]
When I run this command:

```
composer install
```

I get the following output:

```
- Installing google/apiclient-services (dev-master e049d8c): Downloading (100%)
    Failed to download google/apiclient-services from dist: Could not delete /home/vagrant/Code/LocalTracSeo/vendor/google/apiclient-services/.nfs0000000000014fb500000001:
    Now trying to download from source

```

This problem seem to be cause because something, in composer, tries to delete everyfile within the folder. NFS (used to share files accross network), put files with the format .nfsXXX to mark folder to be deleted.

Composer should not attempt to delete these system file AND should wait a couple of microsecond so that the files get correctly delete over network. I tried to change the nfs to be async, but unfortunately didn't work.

**Comment by Seldaek** [comment]
Which OS/platform are you experiencing this on?

**Comment by loiclavoie** [comment]
Plateform Host: Windows 7
VM Host : Vagrant Homestead ( Ubuntu 16.04.2 ).
Vagrant plugins: vagrant-winnfsd

Folder shared from Windows 7 throught NFS Server with NFS enabled, configured throught the normal NFS shared_folder of vagrant.

Note that this issue happen only during certain module. Somehow, I think the execution may happen too fast and some post-script of composer attempt some "rm" somehow (hypothesis, I haven't check the composer code). When this rm happen, it attempt to remove a system file (.nfs) which cannot be touch since it's a file used to tell async nfs to "delete this folder on the remote host".

It happened to me with the composer package "google/apiclient-services"

**Comment by Seldaek** [comment]
Is there a chance you try a different virtual filesystem / sync method to share files with the VM?

**Comment by loiclavoie** [comment]
@Seldaek Yes and no. I mean, at home, I have a Hyper-V based plateform on top on windows 10 (with each VM being either RedHat or Centos 7) and I never had that issue. Though I use cifs to share folder from the host (Win10) to the VM (Centos7).

At work, I'm currently limited to the Win7 install they gave me. When I get my new computer, I'll have Win10 and most likely won't rely on VirtualHost (oracle) and will go for HyperV, so I'll test then with a new share. 

Do you know if anywhere in Composer there's a attempt to delete AFTER downloading the dependancy?
If so, the quick workaround solution might be to leave a timed validation of 50000μ after the rm but before whatever validataion (meaning also that you'll have to catch the system refusal for deleting the folder). Another trick might be to exclude any file starting with dot nfs (.nfs). That way, you assume that the rm has correctly function and go forward with your script.

**Comment by Seldaek** [comment]
Yeah the problem is sometimes on windows rmdir fails because the dir is locked, so we do retry already in that case after waiting 50ms, but I guess in this case maybe the exec of rm command goes through but then it checks if the file exists and it's still there so it tries to delete it recursively with php and at that point it goes over the .nfs files.. Not too sure what we can reasonably do here really.

As for deleting after downloading, it actually makes sure the vendor/foo/bar folder is present but empty, not sure if that's the exact step at which it fails.. You'd have to investigate more in the code.

**Comment by loiclavoie** [comment]
Well, isn't 50ms rather short? 
It extremely long for a "local" folder, but any network folder or share might end up causing problem (like in the case of cifs or NFS).
From my understanding (I'm not a filesystem expert), sync-logic network share sends the command of removing a file (to the host) and then ask for a confirmation that the file has been remotely deleted before actually deleting it on the client server. Assuming normal network latency, I think we could easilly increase that 50ms to a 500 or 750ms.

Maybe the code could detect wether or not your file system is local or shared? That way, the 50ms would be a on-exec composer configuration and not a per dependancy variable? (Just a suggestion, throwing spagetti at the wall and seeing what sticks :P )

**Comment by Seldaek** [comment]
Maybe the code could do many things, but to be honest I am not extremely 
keen on adding such workarounds and making it all even more complex to 
support a use case which is almost inexistent.. Seeing as you are the 
first to report an issue in years, and that installing vendors on a 
networked drive isn't generally speaking a common thing to do.

Working on the local drive or using a virtual filesystem that doesn't do 
crazy things sounds like better alternatives to me..

As for the timing issue, I would expect that an unlink() on a 
network-shared file would just block until the thing has been done. How 
the virtual file system handles that and how long it takes doesn't 
matter, but releasing it immediately as "done" while the files are still 
there is really messed up.

**Comment by avant1** [comment]
I've faced similar issue too when using Vagrant on Windows with `winnfsd` Vagrant plugin.
The workarounds I used were:
- rerun `composer install` several times, this error is not repeated all the time
- install dependencies from host (Windows), probably with `--ignore-platform-reqs` flag
- the stupid one: copy `composer.json` and `composer.lock` outside of shared folder, run `composer install` there, copy `vendor/` back to shared folder

I don't think that something should be fixed in Composer.
There are similar issues with other packages (Symfony `cache:clear` may not work, cache in dev mode may not be updated, Twig templates may not be recompiled). It doesn'tmean that Composer, Symfony, Twig or any package should be "fixed". It means that NFS support for Vagrant on Windows sometimes is unreliable. But there are workarounds that allow to mitigate issues related to Composer.

**Comment by nicolasbadia** [comment]
I also had a lot of issues with composer lately and I was using vagrant `1.9.3`. A fix related to NFS and linux has been published in `1.9.4`. I upgraded to `1.9.5` and I do not have issues anymore!

For the record, the issues I had were mostly: `Failed to decode response: zlib_decode(): data error`

**Comment by jni-heyloyalty** [comment]
I have the same issue when running composer on a shared NFS folder via vagrant. 
I don't think it is a composer issue, but rather a Windows issue as file sharing in Windows does not always work "as expected" and pipes are often left open even they have been idle for minutes.

However doing the doing the stupid one works and it works faster:  
- copy `composer.json` and `composer.lock` outside of shared folder, 
- run `composer install` there, 
- copy `vendor/` back to shared folder

Be aware that generating autoloading files might fail if your composer.json refers to classes outside of the vendor lib, which it typically does. Remember to copy yours composer.* files (especially composer.lock) back to the origin place if you ran a 'composer update' instead of 'composer install'.

**Comment by Seldaek** [comment]
Is this still an actual issue? Is there an easy way we can reproduce this?

**Comment by mtibben** [comment]
Ran into this issue today using an NFS share

From http://nfs.sourceforge.net/#faq_d2
> D2. What is a "silly rename"? Why do these .nfsXXXXX files keep showing up?
> A. Unix applications often open a scratch file and then unlink it. They do this so that the file is not visible in the file system name space to any other applications, and so that the system will automatically clean up (delete) the file when the application exits. This is known as "delete on last close", and is a tradition among Unix applications.
>
> Because of the design of the NFS protocol, there is no way for a file to be deleted from the name space but still remain in use by an application. Thus NFS clients have to emulate this using what already exists in the protocol. If an open file is unlinked, an NFS client renames it to a special name that looks like ".nfsXXXXX". This "hides" the file while it remains in use. This is known as a "silly rename." Note that NFS servers have nothing to do with this behavior.
>
> After all applications on a client have closed the silly-renamed file, the client automatically finishes the unlink by deleting the file on the server. Generally this is effective, but if the client crashes before the file is removed, it will leave the .nfsXXXXX file. If you are sure that the applications using these files are no longer running, it is safe to delete these files manually.
>
> The NFS version 4 protocol is stateful, and could actually support delete-on-last-close. Unfortunately there isn't an easy way to do this and remain backwards-compatible with version 2 and 3 accessors.
